com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/StageAppStep.java

```diff
@@ -1,68 +1,94 @@
-        return Arrays.asList(new PollStageAppStatusExecution(recentLogsRetriever));
-    }
-
-    @Override
-    public Integer getTimeout(DelegateExecution context) {
-        return StepsUtil.getVariableOrDefault(context, Constants.PARAM_START_TIMEOUT, Constants.DEFAULT_START_TIMEOUT);
-    }
-
-}
+package com.sap.cloud.lm.sl.cf.process.steps;
+
+import java.util.Arrays;
+import java.util.List;
+
+import javax.inject.Inject;
+
+import org.cloudfoundry.client.lib.CloudControllerClient;
+import org.cloudfoundry.client.lib.CloudControllerException;
+import org.cloudfoundry.client.lib.CloudOperationException;
+import org.cloudfoundry.client.lib.StartingInfo;
+import org.cloudfoundry.client.lib.domain.CloudApplication;
+import org.flowable.engine.delegate.DelegateExecution;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.config.BeanDefinition;
+import org.springframework.context.annotation.Scope;
+import org.springframework.stereotype.Component;
+
+import com.sap.cloud.lm.sl.cf.client.XsCloudControllerClient;
+import com.sap.cloud.lm.sl.cf.core.cf.PlatformType;
+import com.sap.cloud.lm.sl.cf.core.cf.clients.RecentLogsRetriever;
+import com.sap.cloud.lm.sl.cf.core.util.ApplicationConfiguration;
+import com.sap.cloud.lm.sl.cf.process.Constants;
+import com.sap.cloud.lm.sl.cf.process.message.Messages;
+
+@Component("stageAppStep")
+@Scope(BeanDefinition.SCOPE_PROTOTYPE)
+public class StageAppStep extends TimeoutAsyncFlowableStep {
+
+    @Inject
+    protected RecentLogsRetriever recentLogsRetriever;
+
+    @Override
+    protected String getIndexVariable() {
+        return Constants.VAR_APPS_INDEX;
+    }
+
+    @Autowired
+    private ApplicationConfiguration configuration;
+
+    @Override
+    protected StepPhase executeAsyncStep(ExecutionWrapper execution) {
+        CloudApplication app = StepsUtil.getApp(execution.getContext());
+        try {
+            if (configuration.getPlatformType() == PlatformType.CF) {
+                return stageAppCF(execution, execution.getControllerClient(), app);
+            }
+            
+            XsCloudControllerClient xsClient = execution.getXsControllerClient();
+                
+            return stageApp(execution.getContext(), xsClient, app);
+        } catch (CloudOperationException coe) {
+            CloudControllerException e = new CloudControllerException(coe);
+            getStepLogger().error(e, Messages.ERROR_STAGING_APP_1, app.getName());
+            throw e;
+        }
+    }
+
+    private StepPhase stageAppCF(ExecutionWrapper execution, CloudControllerClient client, CloudApplication app) {
+        String uploadToken = (String) execution.getContext()
+            .getVariable(Constants.VAR_UPLOAD_TOKEN);
+        if (uploadToken == null) {
+            return StepPhase.DONE;
+        }
+        String packageGuid = uploadToken.substring(uploadToken.lastIndexOf("/") + 1);

```

Comment:
Naaah, this is not a good idea. Make a more structured UploadToken if necessary, but don't use substring(lastIndexOf()).